hail/src/main/scala/is/hail/expr/ir/MatrixWriter.scala

```diff
@@ -287,44 +330,25 @@ case class SplitPartitionNativeWriter(
           cb.assign(lastSeenSettable, IEmitCode.present(cb, key.copyToRegion(cb, lastSeenRegion, lastSeenSettable.st)))
         }
 
-        cb += ob1.writeByte(1.asInstanceOf[Byte])
-
-        spec1.encodedType.buildEncoder(row.st, cb.emb.ecb)
-          .apply(cb, row, ob1)
-
-        cb += ob2.writeByte(1.asInstanceOf[Byte])
+        buffers.zip(specs).foreach { case (buff, spec) =>
+          cb += buff.writeByte(1.asInstanceOf[Byte])
+          spec.encodedType.buildEncoder(row.st, cb.emb.ecb).apply(cb, row, buff)
+        }
 
-        spec2.encodedType.buildEncoder(row.st, cb.emb.ecb)
-          .apply(cb, row, ob2)
-        cb.assign(n, n + 1L)
+        cb.assign(pCount, pCount + 1L)
       }
 
-      cb.assign(filename1, pctx.asString.loadString(cb))
-      if (hasIndex) {
-        val indexFile = cb.newLocal[String]("indexFile")
-        cb.assign(indexFile, const(index.get._1).concat(filename1).concat(".idx"))
-        indexWriter.init(cb, indexFile)
-      }
-      cb.assign(filename2, const(partPrefix2).concat(filename1))
-      cb.assign(filename1, const(partPrefix1).concat(filename1))
-      cb.assign(os1, Code.newInstance[ByteTrackingOutputStream, OutputStream](mb.createUnbuffered(filename1)))
-      cb.assign(os2, Code.newInstance[ByteTrackingOutputStream, OutputStream](mb.createUnbuffered(filename2)))
-      cb.assign(ob1, spec1.buildCodeOutputBuffer(Code.checkcast[OutputStream](os1)))
-      cb.assign(ob2, spec2.buildCodeOutputBuffer(Code.checkcast[OutputStream](os2)))
-      cb.assign(n, 0L)
+      writeIndexInfo.foreach(_._3.close(cb))
 
-      stream.memoryManagedConsume(region, cb) { cb =>
-        writeFile(cb, stream.element)
+      buffers.foreach { buff =>
+        cb += buff.writeByte(0.asInstanceOf[Byte])
+        cb += buff.flush()
+        cb += buff.close()
       }
 
-      cb += ob1.writeByte(0.asInstanceOf[Byte])
-      cb += ob2.writeByte(0.asInstanceOf[Byte])
-      if (hasIndex)
-        indexWriter.close(cb)
-      cb += ob1.flush()
-      cb += ob2.flush()
-      cb += os1.invoke[Unit]("close")
-      cb += os2.invoke[Unit]("close")
```

Comment:
There's a slight difference in ordering with respect to the index writer. I didn't think that mattered as the the output buffers and index writer seem to be quite separate.